tool-openssl/x509_test.cc

```diff
@@ -47,59 +48,336 @@ X509* CreateAndSignX509Certificate() {
+  ASSERT_TRUE(EVP_PKEY_assign_RSA(pkey.get(), rsa.release()));
+
+  {
+    ScopedFILE signkey_file(fopen(signkey_path, "wb"));
+    ASSERT_TRUE(signkey_file);
+    ASSERT_TRUE(PEM_write_PrivateKey(signkey_file.get(), pkey.get(), nullptr, nullptr, 0, nullptr, nullptr));
+  }
+
+  bssl::UniquePtr<X509_REQ> req(X509_REQ_new());
+  ASSERT_TRUE(req);
+  X509_REQ_set_pubkey(req.get(), pkey.get());
+  X509_REQ_sign(req.get(), pkey.get(), EVP_sha256());
+
+  {
+    ScopedFILE in_file(fopen(in_path, "wb"));
+    ASSERT_TRUE(in_file);
+    ASSERT_TRUE(PEM_write_X509_REQ(in_file.get(), req.get()));
+  }
+
+  args_list_t args = {"-in", in_path, "-out", out_path, "-req", "-signkey", signkey_path};
+  bool result = X509Tool(args);
+  ASSERT_TRUE(result);
+
+  RemoveFile(in_path);
+  RemoveFile(out_path);
+  RemoveFile(signkey_path);
+}
+
+// Test -checkend
+TEST(X509Test, X509ToolCheckEndTest) {
+  char in_path[PATH_MAX];
+
+  ASSERT_GT(createTempFILEpath(in_path), 0u);
+
+  bssl::UniquePtr<X509> x509(CreateAndSignX509Certificate());
+  ASSERT_TRUE(x509);
+
+  {
+    ScopedFILE in_file(fopen(in_path, "wb"));
+    ASSERT_TRUE(in_file);
+    ASSERT_TRUE(PEM_write_X509(in_file.get(), x509.get()));
+  }
+
+  args_list_t args = {"-in", in_path, "-checkend", "3600"};
+  bool result = X509Tool(args);
+  ASSERT_TRUE(result);
+
+  RemoveFile(in_path);
+}
+
+// Test mutually exclusive options, required options, and required arugments
+TEST(X509Test, MutuallyExclusiveOptionsTest) {
+  char in_path[PATH_MAX];
+  char out_path[PATH_MAX];
+  char signkey_path[PATH_MAX];
+
+  ASSERT_GT(createTempFILEpath(in_path), 0u);
+  ASSERT_GT(createTempFILEpath(out_path), 0u);
+  ASSERT_GT(createTempFILEpath(signkey_path), 0u);
+
+  bssl::UniquePtr<EVP_PKEY> pkey(EVP_PKEY_new());
+  ASSERT_TRUE(pkey);
+  bssl::UniquePtr<RSA> rsa(RSA_new());
+  ASSERT_TRUE(rsa);
+  bssl::UniquePtr<BIGNUM> bn(BN_new());
+  ASSERT_TRUE(bn && BN_set_word(bn.get(), RSA_F4) && RSA_generate_key_ex(rsa.get(), 2048, bn.get(), nullptr));
+  ASSERT_TRUE(EVP_PKEY_assign_RSA(pkey.get(), rsa.release()));
+
+  {
+    ScopedFILE signkey_file(fopen(signkey_path, "wb"));
+    ASSERT_TRUE(signkey_file);
+    ASSERT_TRUE(PEM_write_PrivateKey(signkey_file.get(), pkey.get(), nullptr, nullptr, 0, nullptr, nullptr));
+  }
+
+  bssl::UniquePtr<X509> x509(CreateAndSignX509Certificate());
+  ASSERT_TRUE(x509);
+
+  {
+    ScopedFILE in_file(fopen(in_path, "wb"));
+    ASSERT_TRUE(in_file);
+    ASSERT_TRUE(PEM_write_X509(in_file.get(), x509.get()));
+  }
+
+  // -noout with -out, -modulues, -dates, -checkend
+  args_list_t args1 = {"-in", in_path, "-noout", "-out", out_path};
+  ASSERT_FALSE(X509Tool(args1));
+  args_list_t args2 = {"-in", in_path, "-noout", "-modulus"};
+  ASSERT_FALSE(X509Tool(args2));
+  args_list_t args3 = {"-in", in_path, "-noout", "-dates"};
+  ASSERT_FALSE(X509Tool(args3));
+  args_list_t args4 = {"-in", in_path, "-noout", "-checkend", "3600"};
```

Comment:
Nit: Instead of creating a new variable for each, can we just directly reassign `args` with the new value?
```suggestion
  args = {"-in", in_path, "-noout", "-modulus"};
  ASSERT_FALSE(X509Tool(args));
  args = {"-in", in_path, "-noout", "-dates"};
  ASSERT_FALSE(X509Tool(args));
  args = {"-in", in_path, "-noout", "-checkend", "3600"};
```